rep_cls.py

```diff
@@ -0,0 +1,366 @@
+        file names. Raise ValueError if specified directory does not exist.
+        Also set self._label_img_map."""
+        # Check directory exists
+        if not os.path.isdir(img_dir):
+            raise ValueError('Directory {} does not exist.'.format(img_dir))
+        # Read all image names
+        self._img_names = [os.path.join(img_dir, name) for name in
+                            os.listdir(img_dir)]
+        self._img_dir = img_dir
+
+        # Create map form label to image name
+        _codepoints = [name.split('/')[-1].split('_')[0] for name in
+                       self._img_names]
+        _keys = [chr(int('0x' + codepoint[2:], 16)) for codepoint in
+                 _codepoints]
+        self._label_img_map = dict(zip(_keys, self._img_names))
+
+    @n_candidates.setter
+    def n_candidates(self, n_candidates):
+        """Set self._n_candidates. Raise ValueError if n_candidates is less or
+        equal to zero or is not an integer."""
+        if n_candidates <= 0 or type(n_candidates) != int:
+            raise ValueError("Number of candidates must be a positive integer.")
+        self._n_candidates = n_candidates
+
+    @pca_dimensions.setter
+    def pca_dimensions(self, pca_dimensions):
+        """Create d PCA models for the d dimensions in pca_dmensions. Fit all
+        models with the full embeddings and create reduced embeddings in
+        self._reps. If any element in pca_dimensions is less or equal to 0 or
+        is not an integer, raise ValueError."""
+        for dimension in pca_dimensions:
+            if dimension <= 0 or type(dimension) != int:
+                raise ValueError('PCA component number must be a positive '
+                                 'integer.')
+        self._pca_dimensions = pca_dimensions
+
+        # Build PCA models
+        self._pca_models = []
+        print("Building PCA models.")
+        for dimension in pca_dimensions:
+            self._pca_models.append(PCA(n_components=dimension))
+        # Fit models with full embeddings
+        print("Fitting PCA models.")
+        for model in self._pca_models:
+            model.fit(self.embeddings)
+
+        # Get reduced embeddings as representations
+        print("Generating reduced embeddings as representations.")
+        self._reps =[]
+        for model in self._pca_models:
+            reduced_embeddings = model.fit_transform(self.embeddings)
+            self._reps.append(reduced_embeddings)
+
+    @img_format.setter
+    def img_format(self, img_format):
+        """Raise TypeError if img_format is not an ImgFormat enum class."""
+        # Distance metrics setup
+        if img_format not in list(ImgFormat):
+            raise TypeError('Expect img_format to be a member of Format class.')
+        self._img_format = img_format
+
+    def get_pca_results(self):
+        """Get all embeddings after PCA selection."""
+        return self._reps
+
+    def get_confusables_for_char(self, char):
```

Comment:
This function currently take a two step process, the first step it generate candidate_pool, the second step it generate confusables. Could we split this into 3 functions, the first one implement the the first part to generate candidate_pool, the second one generate confusables from candidate_pool. and the third one call the first one to get candidate_pool and pass that to the second one to get the confusables.

Reply:
Good idea! That would improve testability too.